src/mongo/mongo.ts

```diff
@@ -30,129 +33,138 @@ export interface IMongoResource extends vscode.TreeItem {
-				});
-		});
+	async getChildren(): Promise<IMongoResource[]> {
+		let azureServers = this._azureServers;
+
+		if (this._isLoading || this.azureAccount.status === "Initializing" || this.azureAccount.status === "LoggingIn") {
+			azureServers = [new LoadingNode()];
+		} else if (this.azureAccount.status === "LoggedOut") {
+			azureServers = [new SignInToAzureNode()];
+		} else if (azureServers.length === 0) {
+			azureServers = [new AddResourceFilterNode()];
+		}
+
+		return azureServers.concat(this._servers);
 	}
 
-	add(connectionString: string) {
-		this.resolveServer(connectionString, true)
-			.then(server => {
-				this._serverConnections.push(connectionString);
-				this._serversJson.write(this._serverConnections)
-					.then(() => {
-						this._onChange.fire();
-					});
-			});
+	async add(connectionString: string) {
+		try {
+			const db = await MongoClient.connect(connectionString);
+			const server = new Server(connectionString, db.serverConfig);
+			if (this._servers.find(s => s.id === server.id)) {
+				vscode.window.showWarningMessage(`Server '${server.id}' is already connected.`)
+			} else {
+				this._servers.push(server);
+				this._onChange.fire();
+			}
+		} catch (error) {
+			vscode.window.showErrorMessage(error.message);
+		}
 	}
 
 	remove(server: IMongoResource) {
-		const id = server instanceof Server ? server.id : server instanceof NoConnectionServer ? server.id : null;
-		const index = this._servers.findIndex((value) => value.id === id);
+		const index = this._servers.findIndex((value) => value.id === server.id);
 		if (index !== -1) {
 			this._servers.splice(index, 1);
-			this._serversJson.write(this._servers.map(server => server.id));
 			this._onChange.fire();
 		}
 	}
 
-	private resolveServer(connectionString: string, throwError: boolean): Promise<IMongoResource> {
-		return new Promise((c, e) => {
-			MongoClient.connect(connectionString, (error: MongoError, db: Db) => {
-				if (error) {
-					vscode.window.showErrorMessage(error.message);
-					if (throwError) {
-						e(error.message);
-					} else {
-						c(new NoConnectionServer(connectionString, error.message));
-					}
-				} else {
-					c(new Server(connectionString, db.serverConfig));
-				}
-			});
-		})
+	async refreshAzureResources(): Promise<void> {
+		if (!this._isLoading) {
+			this._isLoading = true;
+			try {
+				this._onChange.fire();
+				this._azureServers = await this.getAzureMongoResources();
+			} finally {
+				this._isLoading = false;
+				this._onChange.fire();
+			}
+		}
 	}
-}
 
-export class NoConnectionServer implements IMongoResource {
+	private async getAzureMongoResources(): Promise<IMongoResource[]> {
+		let servers: Server[] = [];
+
+		await Promise.all(this.azureAccount.filters.map(async (filter) => {
+			const docDBClient = new DocumentdbManagementClient(filter.session.credentials, filter.subscription.subscriptionId);
+
+			const serverResult = await Promise.all(filter.resourceGroups.map(async group => {
```

Comment:
We would have to do extensive testing to find out. On one hand it might be more requests this way, but it can also do more asynchronously.

In addition, the 'DatabaseAccount' objects returned by the 'list' operation do not have the resource group name as a part of the metadata. It's almost easier to do it this way so that we know the group name associated with each resource.